File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/SharedViewModelCharaFactory.java
package com.github.malitsplus.shizurunotes.ui;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import java.lang.reflect.InvocationTargetException;

public class SharedViewModelCharaFactory implements ViewModelProvider.Factory {
    private SharedViewModelChara sharedViewModelChara;

    public SharedViewModelCharaFactory(SharedViewModelChara sharedViewModelChara){
        this.sharedViewModelChara = sharedViewModelChara;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        try {
            return modelClass.getConstructor(SharedViewModelChara.class).newInstance(sharedViewModelChara);
        } catch (NoSuchMethodException | IllegalAccessException | InstantiationException | InvocationTargetException e) {
            throw new RuntimeException("Cannot create an instance of " + modelClass, e);
        }
    }
}

File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/shared/SharedViewModelCharaFactory.kt
package com.github.malitsplus.shizurunotes.ui.shared

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import java.lang.reflect.InvocationTargetException

class SharedViewModelCharaFactory(
    private val sharedViewModelChara: SharedViewModelChara
) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return try {
            modelClass.getConstructor(SharedViewModelChara::class.java)
                .newInstance(sharedViewModelChara)
        } catch (e: NoSuchMethodException) {
            throw RuntimeException("Cannot create an instance of $modelClass", e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException("Cannot create an instance of $modelClass", e)
        } catch (e: InstantiationException) {
            throw RuntimeException("Cannot create an instance of $modelClass", e)
        } catch (e: InvocationTargetException) {
            throw RuntimeException("Cannot create an instance of $modelClass", e)
        }
    }
}